### DISCUSSION TITLE
What generics to give to EntityRepository for a Custom Base Repository

### DISCUSSION THREAD

**Discussion by noftaly** [mikro-orm/mikro-orm#2418] [Q&A] [answered]
Hello there!

I'm trying to make a custom base repository to have custom pagination methods. I saw that it was possible on the [configuration](https://mikro-orm.io/docs/configuration/#custom-repository) and [repositories](https://mikro-orm.io/docs/configuration/#custom-repository) pages, but I couldn't find any examples, be it on the docs or in the example GH repos.

So my question is: how can I extend the class `EntityRepository` that requires a generics, if I don't know yet what entity will be managed? Do I just put my custom `BaseEntity` class?

It would be great to have a more explicit documentation about that or more examples, as it is currently just a little line at the end of the page, although I think it is an important feature 😄

Thanks for your answers :]

**Comment by B4nan** [comment, marked as answer]
You will need a generic type argument as well:

```ts
import { EntityRepository } from '@mikro-orm/postgresql'; // use the driver package

export class MyEntityRepository<T> extends EntityRepository<T> { ... }
```

**Reply by noftaly** [reply]
How did I not though of that... Thanks!

**Comment by drackp2m** [comment]
> You will need a generic type argument as well:
> 
> ```ts
> import { EntityRepository } from '@mikro-orm/postgresql'; // use the driver package
> 
> export class MyEntityRepository<T> extends EntityRepository<T> { ... }
> ```

Good morning [Martin](https://github.com/B4nan),

I got something like what you are proposing to work for me, but I don't like that my "MyEntityRepository" has all the methods of "EntityRepository".

What I think complicates things for me, is that the entityName attribute of EntityRepository is protected.

I don't know if anyone else has encountered this problem and can come up with an elegant solution.


Edit:
I have tried something like...

```typescript
export abstract class BaseRepository<T extends AnyEntity<T>> {
	constructor(private readonly repository: EntityRepository<T>) {}

	async getOne(query: FilterQuery<T>): Promise<T | null> {
		const user = await this.repository.findOne(query);

		if (!user) {
			throw new NotFoundException('user not exists');
		}

		return user;
	}

	· · ·
}
```

...and I get the following error: MetadataError: Metadata for entity undefined not found

**Reply by B4nan** [reply]
I don't understand how having a protected property in the class can complicate anything - you will need to have such one in your custom class as well if you want it to operate on multiple entity types.

> I have tried something like...

And where in that code are you saying the repository should work with `User` entity?

**Reply by B4nan** [reply]
You can of course have a service that will just inject something and work with that instead of extending the repository class from the ORM - in that case, you shouldn't import any repository, but the EntityManager, and work with that. And as said above - you still need to tell it what entity type you want to work with.

**Reply by drackp2m** [reply]
Ok, I guess I was mistaken in believing that having the `entityName` property publicly available in `EntityRepository` could help me at all.

However, I have tried to apply the solution you propose, with success, although without quite understanding why it works.

<details>
  <summary>custom-entity.repository.ts</summary>

  ```typescript
  import { EntityManager, FilterQuery, Primary } from '@mikro-orm/core';
  
  import { NotFoundException } from '../exception/not-found.exception';
  
  import { CustomBaseEntity } from './base.entity';
  
  export class CustomEntityRepository<T extends CustomBaseEntity<T>> {
	  constructor(
		  private readonly entityManager: EntityManager,
		  private readonly entityName: string,
	  ) {}
  
	  getReference(id: Primary<T>): T {
		  return this.entityManager.getReference(this.entityName, id);
	  }
  
	  async getOne(query: FilterQuery<T>): Promise<T | null> {
		  const user = await this.entityManager.findOne(this.entityName, query);
  
		  if (!user) {
			  const entityName = this.entityName.toString().replace('Entity', '').toLocaleLowerCase();
			  throw new NotFoundException('not exists', entityName);
		  }
  
		  return user;
	  }
  
	  async getMany(): Promise<T[]> {
		  return this.entityManager.find(this.entityName, {});
	  }
  
	  ···
  }
  ```
</details>

What I don't understand is why I can have my `UserRepository` extending from `CustomEntityRepository`, without ever instantiating this one, passing it `EntityManager` and `entityName`.


<details>
  <summary>user.repository.ts</summary>

  ```typescript
  import { Injectable } from '@nestjs/common';
  
  import { CustomEntityRepository } from '../../shared/util/custom-entity.repository';
  
  import { UserEntity } from './user.entity';
  
  @Injectable()
  export class UserRepository extends CustomEntityRepository<UserEntity> {
	  findActiveUsers(): UserEntity[] {
		  return [];
	  }
  }
  ```
</details>

Then I also have my entity:

```typescript
@Entity({ tableName: 'users', customRepository: () => UserRepository })
@ObjectType({ description: 'user' })
export class UserEntity extends CustomBaseEntity<UserEntity> {
	@Property({ unique: true })
	@Field()
	username!: string;
        
        ···
```

And then I import my `UserRepository` where I need it and everything works "magically".
No need to define at any time `EntityManager` or `entityName`, I imagine that the first one comes automatically when importing `MikroOrmModule` in `AppModule`, but the entity name...

**Comment by superiums** [comment]
if you're using folder based entity discovery, `@Entity({abstract:true})` is needed.

**Comment by drackp2m** [comment]
> if you're using folder based entity discovery, `@Entity({abstract:true})` is needed.

Thanks for your comment @superiums, the problem is that I am using "folder-based discovery" (half-heartedly).

Let me explain: in a nutshell, I'm using Nx to set up a mono-repo with NestJS + Angular, and in entity management I'm using MikroORM.

The problem comes, in that Nx compiles all the code of my NestJS application in a single index.js file in the directory `dist/apps/{project_name}/main.js`.

And being so, I never knew how to get it to extract the JS entities through a regular expression, because they are all inside that one file, but there are many more things mixed in there, so, the only thing I could think of, is to load them "manually" from my TS source code, using the [following method](https://github.com/drackp2m/set-online/blob/main/apps/api/src/shared/module/config/mikro-orm.config.ts#L22).

It's not pretty, but it works in both NestJS and the MikroORM CLI (migrations).

Likewise, my "base-entity" from which all the others extend, does have [this decorator](https://github.com/drackp2m/set-online/blob/main/apps/api/src/shared/util/base.entity.ts#L13) you mention.

**Reply by superiums** [reply]
maybe [following method](https://github.com/drackp2m/set-online/blob/main/apps/api/src/shared/module/config/mikro-orm.config.ts#L22) is good.

i preffere another way, in nest , we can turn on the auto discovery. and then inject entity by `mikroorm.forFeather(myEntity)` in every module.
and in CLI, we can use config : `entityTs: [paths].`

no matter which way you use, have the decorator is a good habit.

**Reply by drackp2m** [reply]
I liked this option, but I have several problems with it.

Apparently, the new `defineConfig` method, imported from `'@mikro-orm/postgresql'`, does not support the `autoLoadEntities` attribute. I don't know if this is a bug, but....

```
@deprecated - type option will be removed in v6, use defineConfig exported from the driver package to define your ORM config
```

Anyway, I started to read the documentation (again) and I saw that there is another way besides the two I knew (use type: 'postgresql' or use the defineConfig method), and that is to use `driver`. https://mikro-orm.io/docs/configuration#driver

---

Anyway, right after solving that, comes the misfortune (it always gave me a lot of trouble trying to configure everything right).

My configuration was like this, and everything was fine:

<details>
    <summary>mikro-orm.config.ts</summary>

```ts
export default async (): Promise<MikroOrmModuleSyncOptions> =>
	defineConfig({
		...databaseConfig(),
		allowGlobalContext: true,
		entities: await getEntities(),
		entitiesTs: ['apps/api/src/module/**/*.entity.ts'],
		forceUtcTimezone: true,
		migrations: {
			tableName: 'migrations',
			path: 'apps/api/migrations',
			snapshot: false,
			silent: true,
		},
	});
```
</details>


<details>
    <summary>Debug over NestJS start process</summary>

```
[info] MikroORM version: 5.7.14
[discovery] ORM entity discovery started, using ReflectMetadataProvider
[discovery] - processing entity UserEntity
[discovery] - processing entity CustomBaseEntity
[discovery] - processing entity BaseEntity
[discovery] - entity discovery finished, found 3 entities, took 24 ms
[info] MikroORM successfully connected to database set on postgresql://set:*****@set-db:5432
[query] select 1 from pg_database where datname = 'set' [took 1 ms, 1 result]
```
</details>



<details>
    <summary>Debug over mikro-orm cli</summary>

```
❯ mikro-orm debug
Current MikroORM CLI configuration
 - dependencies:
   - mikro-orm 5.7.14
   - node 20.2.0
   - knex 2.5.1
   - pg 8.11.1
   - typescript 5.1.6
 - package.json found
 - ts-node enabled
 - searched config paths:
   - /usr/src/app/apps/api/src/shared/module/config/mikro-orm.config.ts (found)
   - /usr/src/app/src/mikro-orm.config.ts (not found)
   - /usr/src/app/mikro-orm.config.ts (not found)
   - /usr/src/app/dist/mikro-orm.config.js (not found)
   - /usr/src/app/mikro-orm.config.js (not found)
 - configuration found
 - database connection succesful
 - will use `entities` array (contains 1 references and 0 paths)
```
</details>

However, following your advice, and trying to try one more time (because it has been many times already), I modify my configuration as follows:

<details>
    <summary>mikro-orm.config.ts</summary>

```ts
export default async (): Promise<MikroOrmModuleSyncOptions> => {
	return {
		...databaseConfig(),
		allowGlobalContext: true,
		entitiesTs: ['apps/api/src/module/**/*.entity.ts'],
		driver: PostgreSqlDriver,
		autoLoadEntities: true,
		debug: true,
		forceUtcTimezone: true,
		migrations: {
			tableName: 'migrations',
			path: 'apps/api/migrations',
			snapshot: false,
			silent: true,
		},
		tsNode: true,
	};
};
```
</details>

Now, my NestJS debug looks like this:

<details>
    <summary>Debug over NestJS start process</summary>

```
[info] MikroORM version: 5.7.14
[discovery] ORM entity discovery started, using ReflectMetadataProvider
[discovery] - processing 1 files
[Nest] 6005  - 08/16/2023, 8:34:20 PM   ERROR [ExceptionHandler] Cannot use import statement outside a module
/usr/src/app/apps/api/src/module/user/user.entity.ts:1
import { Entity, Enum, Property } from '@mikro-orm/core';
^^^^^^

SyntaxError: Cannot use import statement outside a module
    at internalCompileFunction (node:internal/vm:73:18)
    at wrapSafe (node:internal/modules/cjs/loader:1175:20)
    at Module._compile (node:internal/modules/cjs/loader:1219:27)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1309:10)
    at Module.load (node:internal/modules/cjs/loader:1113:32)
    at Function.Module._load (node:internal/modules/cjs/loader:960:12)
    at Function.Module._load (/usr/src/app/node_modules/@nx/js/src/executors/node/node-with-require-overrides.js:18:31)
    at Module.require (node:internal/modules/cjs/loader:1137:19)
    at require (node:internal/modules/helpers:121:18)
    at Function.requireFrom (/usr/src/app/node_modules/@mikro-orm/core/utils/Utils.js:759:70)
```
</details>

And MikroORM's also gives some failures...

<details>
    <summary>Debug over mikro-orm cli</summary>

```
❯ mikro-orm debug
Current MikroORM CLI configuration
 - dependencies:
   - mikro-orm 5.7.14
   - node 20.2.0
   - typescript 5.1.6
 - package.json found
 - ts-node enabled
 - searched config paths:
   - /usr/src/app/apps/api/src/shared/module/config/mikro-orm.config.ts (found)
   - /usr/src/app/src/mikro-orm.config.ts (not found)
   - /usr/src/app/mikro-orm.config.ts (not found)
   - /usr/src/app/dist/mikro-orm.config.js (not found)
   - /usr/src/app/mikro-orm.config.js (not found)
- configuration not found (No entities found, please use `entities` option)
```
</details>

Thus, neither NestJS starts, nor can I execute commands like `mikro-orm migration:pending`:

<details>
    <summary>$ mikro-orm migration:pending</summary>

```
❯ mikro-orm migration:pending
mikro-orm migration:pending

List all pending migrations

Options:
  -v, --version  Show version number                                   [boolean]
  -h, --help     Show help                                             [boolean]

Error: No entities found, please use `entities` option
    at Configuration.validateOptions (/usr/src/app/node_modules/@mikro-orm/core/utils/Configuration.js:226:19)
    at new Configuration (/usr/src/app/node_modules/@mikro-orm/core/utils/Configuration.js:29:18)
    at Function.getConfiguration (/usr/src/app/node_modules/@mikro-orm/core/utils/ConfigurationLoader.js:38:24)
    at async Function.getORM (/usr/src/app/node_modules/@mikro-orm/cli/CLIHelper.js:22:25)
    at async Function.handleMigrationCommand (/usr/src/app/node_modules/@mikro-orm/cli/commands/MigrationCommandFactory.js:73:21)
```
</details>

If your configuration really works for you @superiums, I would be very grateful if you could show me how you have configured all this (maybe if your project is public you can pass me the link to the repository).
Since I've been trying for months, on and off, to make this a little nicer, but there is no way.

**Reply by B4nan** [reply]
> Apparently, the new defineConfig method, imported from '@mikro-orm/postgresql', does not support the autoLoadEntities attribute. I don't know if this is a bug, but....

`autoLoadEntities` is not an ORM config option, it is nestjs specific. You can keep it there even with `defineConfig`, but you will need cast the object to `any` or put `@ts-ignore` comment on the problematic like. `defineConfig` will only accept known ORM config options, as otherwise we wouldn't be able to detect typos.

**Reply by drackp2m** [reply]
> `autoLoadEntities` is not an ORM config option, it is nestjs specific. You can keep it there even with `defineConfig`, but you will need cast the object to `any` or put `@ts-ignore` comment on the problematic like. `defineConfig` will only accept known ORM config options, as otherwise we wouldn't be able to detect typos.

Quite right Martin. Also, I looked at the documentation now, which specifically explains that `defineConfig` is intended to provide intellisense even in JavaScript. Although the `driver: PostgreSqlDriver` option exists, I do not understand why the deprecated message does not also show you this alternative.

I also just realized that NestJS was giving me the error `SyntaxError: Cannot use import statement outside a module` because I kept importing my entities through JS transpiles while I had `tsNode: true` in my global configuration.

However, now there is something I still don't quite understand. I left the option of `tsNode: true`, which will affect both NestJS and the CLI, and that being the case, I thought I could bypass the `entieies` configuration and just focus on `entitiesTs`, however, the code fails me if I put `entitiesTs: ['apps/api/src/module/**/*.entity.ts']`, but everything works perfect, if I rename the key to `entities` (even if it points to .ts files). I don't understand anything... What is `entitiesTs` for?

But well, it works, I'm happy...

**Reply by B4nan** [reply]
With `tsNode: true`, the ORM will use `entitiesTs` option instead of `entities`. This is not meant for the CLI, there the TS support is added via `useTsNode` in the package.json (and by installing `ts-node`).

By enabling this explicitly, you override the auto-detection, and `entitiesTs` will be always used, regardless of whether the ORM detects TS support. You don't want that for production, the `tsNode` flag is mostly for tests.

**Reply by drackp2m** [reply]
> By enabling this explicitly, you override the auto-detection, and `entitiesTs` will be always used, regardless of whether the ORM detects TS support. You don't want that for production, the `tsNode` flag is mostly for tests.

When you say I overwrite `auto-detection`, do you mean it stays on by default? In production, I think that mikro-orm with tsNode, I will only need it to run the .ts migrations that are generated during the development of my application.

Currently, these migrations are not imported by any file in the project, so they are not compiled and do not end up in /dist. I have yet to run a script to copy or transpile them into JS when doing the build.

But my main problem is that my Nest application is compiled in a single index.js file, and so it is impossible for me to import entities using directory discovery. So, that's when I found @superiums idea useful.

What I still don't understand, however, is why, if I activate tsNode and leave entities as empty array and entitiesTs pointing to the directory where my TS entities are, this way:

```ts
	tsNode: true,
	entities: [],
	entitiesTs: ['apps/api/src/module/**/*.entity.ts'],
```

I get this error:

```
[Nest] 4584  - 08/18/2023, 6:52:21 PM   ERROR [ExceptionHandler] No entities found, please use `entities` option

Error: No entities found, please use `entities` option
    at Configuration.validateOptions (/usr/src/app/node_modules/@mikro-orm/core/utils/Configuration.js:226:19)
    at new Configuration (/usr/src/app/node_modules/@mikro-orm/core/utils/Configuration.js:29:18)
    at new MikroORM (/usr/src/app/node_modules/@mikro-orm/core/MikroORM.js:54:27)
    at Function.init (/usr/src/app/node_modules/@mikro-orm/core/MikroORM.js:29:21)
    at Injector.instantiateClass (/usr/src/app/node_modules/@nestjs/core/injector/injector.js:369:37)
    at callback (/usr/src/app/node_modules/@nestjs/core/injector/injector.js:65:34)
    at Injector.resolveConstructorParams (/usr/src/app/node_modules/@nestjs/core/injector/injector.js:144:24)
    at Injector.loadInstance (/usr/src/app/node_modules/@nestjs/core/injector/injector.js:70:13)
    at Injector.loadProvider (/usr/src/app/node_modules/@nestjs/core/injector/injector.js:97:9)
    at /usr/src/app/node_modules/@nestjs/core/injector/instance-loader.js:56:13
```

But if I define a directory of entities that does not exist, like:

```ts
	tsNode: true,
	entities: ['not-exists'],
	entitiesTs: ['apps/api/src/module/**/*.entity.ts'],
```

So it does seem to use the content of entitiesTs, but I get this error:

```
[Nest] 5382  - 08/18/2023, 6:55:24 PM   ERROR [ExceptionHandler] Cannot use import statement outside a module
/usr/src/app/apps/api/src/module/user/user.entity.ts:1
import { Entity, Enum, Property } from '@mikro-orm/core';
^^^^^^

SyntaxError: Cannot use import statement outside a module
    at internalCompileFunction (node:internal/vm:73:18)
    at wrapSafe (node:internal/modules/cjs/loader:1175:20)
    at Module._compile (node:internal/modules/cjs/loader:1219:27)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1309:10)
    at Module.load (node:internal/modules/cjs/loader:1113:32)
    at Function.Module._load (node:internal/modules/cjs/loader:960:12)
    at Function.Module._load (/usr/src/app/node_modules/@nx/js/src/executors/node/node-with-require-overrides.js:18:31)
    at Module.require (node:internal/modules/cjs/loader:1137:19)
    at require (node:internal/modules/helpers:121:18)
    at Function.requireFrom (/usr/src/app/node_modules/@mikro-orm/core/utils/Utils.js:759:70)
```

**Comment by superiums** [comment]
in my case:

### .env
```

# mikro
MIKRO_ORM_CLI = ./mikro-orm.config.ts
MIKRO_ORM_CLI_USE_TS_NODE = true
MIKRO_ORM_TYPE = postgresql
# MIKRO_ORM_ENTITIES = ./dist/**/**/entity/*.entity.js
# MIKRO_ORM_ENTITIES_TS = ./apps/**/src/entity/*.entity.ts, ./libs/**/src/entity/*.entity.ts
MIKRO_ORM_DB_NAME = nsdb
MIKRO_ORM_HOST = localhost
MIKRO_ORM_PORT = 5432
MIKRO_ORM_USER = nsuser
MIKRO_ORM_PASSWORD = nesttest
MIKRO_ORM_MIGRATIONS_PATH = ./migrations
MIKRO_ORM_MIGRATIONS_PATTERN = ^[\w-]+\d+\.js$
MIKRO_ORM_POPULATE_AFTER_FLUSH = true
MIKRO_ORM_FORCE_ENTITY_CONSTRUCTOR = true
MIKRO_ORM_FORCE_UNDEFINED = true
MIKRO_ORM_MIGRATIONS_SAFE = true
```
no entities registered here.


### mikroorm-config.ts
```ts

const config = {
    /** used for cli */
    entities: [user , role, org, mo, action,  ], // no need for `entitiesTs` this way
    // entitiesTs: ['./libs/**/src/entity/*.entity.ts', './apps/**/src/entity/*.entity.ts'],
    
    seeder: {
        // path: './seeders', // path to the folder with seeders
        pathTs: './seeders', // path to the folder with TS seeders (if used, we should put path to compiled files in `path`)
        defaultSeeder: 'InitSeeder', // default seeder class name
        glob: '!(*.d).{js,ts}', // how to match seeder files (all .js and .ts files, but not .d.ts)
        emit: 'ts', // seeder generation mode
        fileName: (className: string) => className, // seeder file naming convention
    },
};

export default config;
```

here, i use ts file as entities, but only for CLI.


### app.module.ts
```ts

@Module({
  imports: [
    ConfigModule.forRoot({
      isGlobal: true,
      load: config,
      envFilePath: ['.env'],
    }),
    LoggerModule,
    ...,

    MikroOrmModule.forRoot({
 
      autoLoadEntities: true,                                         //-------------important
      loadStrategy: LoadStrategy.JOINED,
      debug: true,

      subscribers: [new OrgSubscriber(), new RoleSubscriber()]
    }),

  ],
  controllers: [AppController],
  providers: [AppService],
})
export class AppModule implements OnModuleInit {
  onModuleInit() {
    console.log('main Module inited');
  }
}

```
here , we enable `autoLoadEntities`

### xxx.module.ts
```ts
@Module({
  imports: [MikroOrmModule.forFeature([user, role, org])],              //-------------important
  ...
})
export class MembershipModule {
}


```

here, in every module used entity, we register it. 
in this way, every module is good structured, and separated entirely.

**Reply by superiums** [reply]
also , we need to enable tsnode in package.json
```json
"mikro-orm": {
    "useTsNode": true,
    "tsConfigPath": "./tsconfig.json",
    "configPaths": [
      "./mikro-orm.config.ts"
    ]
  },
```

**Reply by drackp2m** [reply]
Thank you very much @superiums, I see that actually our configurations are quite similar.

In your `mikroorm-config.ts` file you manually load, one by one, your entities in the entities array, I discarded this, because I don't want to have to go to this file to add each of my created entities. I don't mind adding them in the modules, since I get their associated repositories automatically injected, at that point I don't mind, because these are files I'm going to have to create anyway and where I'm going to be "working", but I don't want to have to go touch too much the configuration files, like `mikroorm-config.ts`.

Then, I have also seen that your AppModule is much cleaner when importing MikroOrmModule, because in the .env you already load the variables with the name that MikroORM expects to find directly, I load my DB configuration with generic name, like DB_HOST or DB_PORT, and inject the ConfigurationService of NestJS to the forRootAsync method of MikroORM. But the result ends up being the same.

This may also be something I "copy" myself.


And regarding the package.json, since you already define MIKRO_ORM_CLI in your .env, you could remove that key inside your package, likewise, you could replace tsConfigPath by MIKRO_ORM_CLI_TS_CONFIG_PATH and configPaths by MIKRO_ORM_CLI in your .env.

In this way, you will leave your package much cleaner.

Thank you so much for sharing your setup!!!!

**Reply by drackp2m** [reply]
Well, about that last thing I recommended you. I had always defined that configuration in the package.json file, just like you have it now. And seeing that the MikroORM config can be loaded directly into the .env, I thought that the cli config could be loaded as well, but I guess that .env doesn't load "magically" when you just run /node_modules/.bin/mikro-orm, so I added that configuration:

```
      MIKRO_ORM_CLI: apps/api/src/shared/module/config/mikro-orm.config.ts
      MIKRO_ORM_CLI_TS_CONFIG_PATH: apps/api/tsconfig.app.json
      MIKRO_ORM_CLI_USE_TS_NODE: 1
```

To my docker-compose.yml, in the `environment` section.

I have tried to apply my suggestion (after writing the comment) and realized that it is an absolute error. :3
